lib/IDE/Formatting.cpp

```diff
@@ -0,0 +1,859 @@
+size_t getOffsetOfLine(unsigned LineIndex, StringRef Text) {
+  //  SourceLoc start = SourceLoc(llvm::SMLoc::getFromPointer(Text.begin()));
+  // FIXME: We should have a cached line map in EditableTextBuffer, for now
+  // we just do the slow naive thing here.
+  size_t LineOffset = 0;
+  unsigned CurrentLine = 0;
+  while (LineOffset < Text.size() && ++CurrentLine < LineIndex) {
+    LineOffset = Text.find_first_of("\r\n", LineOffset);
+    if (LineOffset != std::string::npos) {
+      ++LineOffset;
+      if (LineOffset < Text.size() &&
+          Text[LineOffset - 1] == '\r' && Text[LineOffset] == '\n')
+        ++LineOffset;
+    }
+
+  }
+  if (LineOffset == std::string::npos)
+    LineOffset = 0;
+  return LineOffset;
+}
+
+size_t getOffsetOfTrimmedLine(unsigned LineIndex, StringRef Text) {
+  size_t LineOffset = getOffsetOfLine(LineIndex, Text);
+
+  // Skip leading whitespace.
+  size_t FirstNonWSOnLine = Text.find_first_not_of(" \t\v\f", LineOffset);
+  if (FirstNonWSOnLine != std::string::npos)
+    LineOffset = FirstNonWSOnLine;
+
+  return LineOffset;
+}
+
+llvm::StringRef getTrimmedTextForLine(unsigned LineIndex, StringRef Text) {
+  size_t LineOffset = getOffsetOfTrimmedLine(LineIndex, Text);
+  size_t LineEnd = Text.find_first_of("\r\n", LineOffset);
+  return Text.slice(LineOffset, LineEnd);
+}
+
+size_t getExpandedIndentForLine(unsigned LineIndex, CodeFormatOptions Options,
+                                                     StringRef Text) {
+  size_t LineOffset = getOffsetOfLine(LineIndex, Text);
+
+  // Tab-expand all leading whitespace
+  size_t FirstNonWSOnLine = Text.find_first_not_of(" \t\v\f", LineOffset);
+  size_t Indent = 0;
+  while (LineOffset < Text.size() && LineOffset < FirstNonWSOnLine) {
+    if (Text[LineOffset++] == '\t')
+      Indent += Options.TabWidth;
+    else
+      Indent += 1;
+  }
+  return Indent;
+}
+
+SwiftEditorLineRange limit(SwiftEditorLineRange Range, SourceManager &SM,
+                                                             unsigned bufferID){
+  auto end = SM.getRangeForBuffer(bufferID).getEnd();
+  auto LineAndColumn = SM.getLineAndColumn(end);
+  auto LastLine = LineAndColumn.first;
+  if (Range.endLine() > LastLine) {
+    return SwiftEditorLineRange(Range.startLine(), LastLine);
+  }
+  return Range;
+}
+
+std::pair<SwiftEditorLineRange, std::string>
+reformat(SwiftEditorLineRange Range, CodeFormatOptions Options,
+                                            SourceManager &SM, SourceFile &SF) {
+
+  auto sourceBufferID = SF.getBufferID().getValue();
+  auto limitedRange = limit(Range, SM, sourceBufferID);
+  FormatWalker walker(SF, SM);
+  CodeFormatter CF(Options);
+  std::string output;
+
+  auto start = limitedRange.startLine();
+  auto length = limitedRange.lineCount();
+  for( auto currentLine : range(start, start + length) ) {
+
+    auto returned = CF.indent(currentLine, walker, SM, sourceBufferID);
+    output.append(returned);
```

Comment:
Please use `raw_string_ostream` instead of repeatedly appending to a `std::string`.